fcrepo-kernel-impl/src/main/java/org/fcrepo/kernel/impl/ContainmentIndexImpl.java

```diff
@@ -87,48 +118,80 @@
+            + " = :transactionId AND " + OPERATION_COLUMN + " = 'delete'";
 
-    private final String IS_CHILD_IN_TRANSACTION = "SELECT TRUE FROM " + TRANSACTION_OPERATIONS_TABLE +
+    private static final String IS_CHILD_ADDED_IN_TRANSACTION = "SELECT TRUE FROM " + TRANSACTION_OPERATIONS_TABLE +
             " WHERE " + FEDORA_ID_COLUMN + " = :child AND " + PARENT_COLUMN + " = :parent" +
             " AND " + TRANSACTION_ID_COLUMN + " = :transactionId AND " + OPERATION_COLUMN + " = 'add'";
 
-    /**
-     * check if a table with the given name exists
-     */
-    private boolean tableIsMissing(final String tableName) throws SQLException {
-        final DatabaseMetaData dbMeta = conn.getMetaData();
-        final ResultSet tables = dbMeta.getTables(null, null, tableName, new String[]{"TABLE"});
-        while (tables.next()) {
-            if (tables.getString("TABLE_NAME").equals(tableName)) {
-                return false;
-            }
-        }
-        return true;
-    }
+    private static final String IS_CHILD_DELETED_IN_TRANSACTION = "SELECT TRUE FROM " + TRANSACTION_OPERATIONS_TABLE +
+            " WHERE " + FEDORA_ID_COLUMN + " = :child AND " + PARENT_COLUMN + " = :parent" +
+            " AND " + TRANSACTION_ID_COLUMN + " = :transactionId AND " + OPERATION_COLUMN + " = 'delete'";
+
+    private static final String ROLLBACK_TRANSACTION = "DELETE FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " +
+            " transactionId = :transactionId";
+
+    private static final String COMMIT_ADD_RECORDS = "INSERT INTO " + RESOURCES_TABLE + " ( " + FEDORA_ID_COLUMN + ", "
+            + PARENT_COLUMN + " ) SELECT " + FEDORA_ID_COLUMN + ", " + PARENT_COLUMN + " FROM " +
+            TRANSACTION_OPERATIONS_TABLE + " WHERE " + TRANSACTION_ID_COLUMN + " = :transactionId AND " +
+            OPERATION_COLUMN + " = 'add'";
+
+    private static final String COMMIT_DELETE_RECORDS = "DELETE FROM " + RESOURCES_TABLE + " WHERE EXISTS(" +
+            " SELECT 1 FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " + RESOURCES_TABLE + "." + FEDORA_ID_COLUMN +
+            " = " + TRANSACTION_OPERATIONS_TABLE + "." + FEDORA_ID_COLUMN + " AND " + RESOURCES_TABLE + "." +
+            PARENT_COLUMN + " = " + TRANSACTION_OPERATIONS_TABLE + "." + PARENT_COLUMN + " AND " +
+            TRANSACTION_OPERATIONS_TABLE + "." + TRANSACTION_ID_COLUMN + " = :transactionId AND " +
+            TRANSACTION_OPERATIONS_TABLE + "." + OPERATION_COLUMN + " = 'delete')";
+
+    private static final String COMMIT_CLEANUP = "DELETE FROM " + TRANSACTION_OPERATIONS_TABLE + " WHERE " +
+            TRANSACTION_ID_COLUMN + " = :transactionId";
 
     /**
      * Connect to the database
      */
     @PostConstruct
-    private void setup() throws SQLException {
-        conn = dataSource.getConnection();
+    private void setup() {
+        final NamedParameterJdbcTemplate jdbcTemplate = getNamedParameterJdbcTemplate();
```

Comment:
JdbcTemplates are thread safe. This class really should only ever have a single instance: https://docs.spring.io/spring/docs/current/spring-framework-reference/data-access.html#jdbc-JdbcTemplate-idioms